### ISSUE TITLE
Model Hallucinates after a few steps when training

### ISSUE THREAD

**Issue by vishwassathish** [EvolvingLMMs-Lab/multimodal-search-r1#19]
Hi there,

I am trying to replicate the results from the paper with the FVQA dataset and the model seems to hallucinate and not use the tools correctly. Here is what I did:

1. Added image search tools with the provided image search result cache in the file `mmsearch_r1/utils/tools/image_search.py`
2. Added text tool with web search API call + JINA reader + Qwen3 summarization into the file `mmsearch_r1/utils/tools/text_search.py`
3. Ran the training with `bash mmsearch_r1/scripts/run_mmsearch_r1_grpo.sh` with the exact hyperparams as the paper (batch size 512, 8 rollouts for GRPO, etc)

Here is the full set of parameters used:

```
python3 -u -m mmsearch_r1.trainer.multimodal.main_ppo \
    algorithm.adv_estimator=grpo \
    data.train_files=~/fvqa_train.parquet \
    data.val_files=~/fvqa_test.parquet \
    data.train_batch_size=512 \
    data.max_prompt_length=4096 \
    data.max_response_length=2048 \
    data.image_key=images \
    data.user_prompt_round_1=mmsearch_r1/prompts/round_1_user_prompt_qwenvl.pkl \
    data.user_prompt_after_image_search=mmsearch_r1/prompts/after_image_search_prompt_qwenvl.pkl \
    data.user_prompt_after_text_search=mmsearch_r1/prompts/after_text_search_prompt_qwenvl.pkl \
    actor_rollout_ref.model.path=Qwen/Qwen2.5-VL-7B-Instruct \
    actor_rollout_ref.actor.optim.lr=2e-6 \
    actor_rollout_ref.actor.optim.lr_sigmoid_decay_warmup=True \
    actor_rollout_ref.actor.optim.lr_sigmoid_decay_ratio=0.95 \
    actor_rollout_ref.actor.optim.lr_sigmoid_decay_warmup_steps=45 \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.ppo_mini_batch_size=128 \
    actor_rollout_ref.actor.entropy_coeff=0 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=8 \
    actor_rollout_ref.actor.use_multi_turn_response_mask=True \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.001 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.fsdp_config.param_offload=False \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=16 \
    actor_rollout_ref.rollout.tensor_model_parallel_size=1 \
    actor_rollout_ref.rollout.name=vllm_multiturn_mmsearch \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.6 \
    actor_rollout_ref.rollout.enable_chunked_prefill=True \
    actor_rollout_ref.rollout.enforce_eager=False \
    actor_rollout_ref.rollout.free_cache_engine=False \
    actor_rollout_ref.rollout.n=8 \
    actor_rollout_ref.rollout.max_gen_round=3 \
    actor_rollout_ref.rollout.response_length_total=8192 \
    actor_rollout_ref.rollout.search.topk=5 \
    actor_rollout_ref.rollout.search.image_search_limit=1 \
    actor_rollout_ref.rollout.search.text_search_limit=2 \
    actor_rollout_ref.rollout.search.parallel_tool_call=True \
    actor_rollout_ref.rollout.search.parallel_tool_call_threads=8 \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=16 \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    algorithm.kl_ctrl.kl_coef=0.001 \
    trainer.critic_warmup=0 \
    trainer.logger=['console','wandb'] \
    trainer.project_name='mmsearch_r1' \
    trainer.experiment_name='testing' \
    trainer.n_gpus_per_node=8 \
    trainer.nnodes=1 \
    trainer.save_freq=20 \
    trainer.test_freq=5 \
    trainer.total_epochs=50 \
    +trainer.search_penalty=0.1 \
    +trainer.format_penalty=0.1 \
    +trainer.reward_mode="SubEM" \
    +trainer.val_before_train=True \
    +algorithm.filter_groups.enable=False 
```

Here is the test accuracy and tool use on wandb

<img width="910" height="638" alt="Image" src="https://github.com/user-attachments/assets/fa8b89bf-96b4-4751-a53d-55825b3f69de" />

The model learns for a while and collapses suddenly. Here is the model output at 20th step

```
<reason>The image shows a person wearing a red dress, but I need to identify the individual. I do not have enough visual information to confirm the identity with certainty.</reason> 

<search><img></search>user 

Searched results: <information>[Image Search Results] The result of the image search consists of web page information related to the image from the user's original question. Each result includes the main image from the web page and its title, ranked in descending order of search relevance, as demonstrated below:1. image:  ............ </information> 

Original user's question: Who is the person wearing the red dress? 
Please extract the visual element relevant to the user's question (such as faces, objects, locations, events, logos, or text) ............. If you decide to give the final answer, place it inside <answer> and </answer>, without detailed explanation or illustration. For example: <answer>Titanic</answer>  

assistant: <reason>The search results confirm that the person in the red dress is Megan Thee Stallion.</reason> 


<answer>Megan Thee Stallion</answer>
```

Upon more investigation, the model seems to hallucinate after 25 steps of training. I have truncated the output to the first ~10% of tokens. The model went on hallucinating.

```
<reason>The building is the Sarajevo Cathedral, which is in Bosnia and Herzegovina.</reason> <answer>Босния и 
Херцеговина</answer</<texttextsouth><<<texttext>><< text text><<<text_text搜索搜索0</<texttext搜<text_search> 
<text_search>text_search>text_search</text_search></text_search>></text搜索></搜索>></text搜索></text搜索></b></b></div></div>></div>
</div></div></div></div></div></div></div></div></d........
```

@ authors did you face this issue? Why is the model hallucinating after training for a while? Please let me know. Thanks!

**Comment by vishwassathish** [comment]
Quick update after some debugging - Seems like the `grad_norms`  go to `inf/nan` around the 20th step:

```
[Global Step: 20] Rollout Ends ...
[Global Step: 20] Reward_fn Starts ...
[Global Step: 20] Reward_fn Ends ...
[Global Step: 20] Compute log_prob Starts ...
[Global Step: 20] Compute log_prob Ends ...
[Global Step: 20] Update Actor Starts ...
[Global Step: 20] Update Actor Ends ...
 WARN: rank 0 grad_norm is not finite: inf
WARN: rank 0 grad_norm is not finite: inf
```

Please let me know if you need more details on my training setup.

**Comment by kim1ng** [comment]
We haven't seen `grad_norm` as `NaN`. The initial erratic outputs you might see are normal; the model is learning the tool format and is corrected over time by a format reward.

For the training crash itself, we've seen in the past that noisy search results (irrelevant or failed responses) may cause sudden behavioral shifts and instability. Could you first check if the search results are clean and relevant?

If they are, we'd recommend lowering the learning rate and seeing if that helps stabilize the training.

**Comment by vishwassathish** [comment]
Thank you for the response. Did you not see `grad_norm` being `inf` either? I corrected some of my text search results. I will comment if the next training run crashes as well.

What I see is that there are certain model outputs that go unchecked by the format rewards. Consider this sequence:

`<reason><<<<reason The building is the Sarajevo Cathedral, which is in Bosnia and Herzegovina.<<<<>></reason> <answer>Sarajevo Cathedral</answer>`

The rewarding is done for a regex match over `r'^<reason>.*</reason>.*<answer>.*</answer>$'` which this string satisfies. If the answer is also correct, there seems to be nothing stopping it from producing erratic outputs.

**Comment by vishwassathish** [comment]
Thank you! There were some noisy text in the output and I managed to remove them. But I still find the issue with gibberish text. Another such output after training for around 50 steps:

`
36m(main_task pid=1827984)[0m [response] [' addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n addCriterion\n<information><p>Inside the search results, the image search tool found the "Welcome to Fabulous Las Vegas" sign at night, confirming that the visual content aligns with the iconic picture of the Las Vegas sign. No further analysis or third-party result is necessary.\n\nBased on this, I have enough information to directly answer the user\'s question.\n\nHence, my response is:\n\n<answer>Las Vegas</answer>\n\n<search><img></search>', 'The search results confirm that the iconic sign is the famous "Welcome to Fabulous Las Vegas" sign. Since the search itself has provided enough visual element to answer the user\'s question, I can confidently provide the answer.\n\nNo further text search is needed.\n\nMy response is:\n\n<answer>Las Vegas</answer>']
`

Was the model finetuned with SFT before the RL finetuning in the paper?

**Comment by vishwassathish** [comment]
I've still not been able to get Qwen 2.5 VL instruct 7B to train successfully with the FVQA dataset and the given code. It seems like the model really does not follow the instructions in `mmsearch_r1/prompts/round_1_user_prompt_qwenvl.pkl`. I see only a few 100 samples (out of 50 steps * 512 batch_size * 8 rollouts ~= 200k sample) using tools correctly to receive the format reward. At batch size of 512, the `grad_norms` also go to `inf`. Eventually the tools end up not being used at all.

Any suggestions would be really helpful. I am using the exact setup given in the code + paper. Also tested with different batch sizes and lowered learning rate as shown below.

<img width="891" height="721" alt="Image" src="https://github.com/user-attachments/assets/7f72c2cd-eefa-4a1e-afa3-520da02ddf56" />

**Comment by kim1ng** [comment]
We directly finetune the model with RL without SFT. I wonder if it might be helpful to try a couple of directions here:
- Reduce the number of search results returned: For example, limiting the tool to top-3 results could reduce noise from external signals. This might make it easier to see whether training becomes more stable with less noise.
- Look into the source of `NaN`: I’ve noticed similar problems reported in verl’s GitHub issues ([#721](https://github.com/volcengine/verl/issues/721), [#891](https://github.com/volcengine/verl/issues/891)). It might be worth checking if the discussions or suggested fixes there line up with what you’re seeing.
- Try to increase the weight of format reward and see if it helps.

**Comment by vishwassathish** [comment]
Thanks! Following the github issues helped a bit. Quick question - is your [verl](https://github.com/volcengine/verl) subdirectory different from the official verl repository? Would there be major differences if I replaced the verl directory with the official one? Asking since [verl](https://github.com/volcengine/verl/commits/main/) has had several improvements since this code was published.

**Comment by kim1ng** [comment]
The `verl` submodule in our repository is based on a specific version ([verl @ 8e9e737](https://github.com/volcengine/verl/tree/8e9e73723fd1cc729bedb3bbcf915060afbda91d)) from several months ago. From what I understand, `verl` has received quite a few updates during this time. You can certainly try to update the submodule in your local repository to the latest version, but please be aware that you might need to resolve some differences.

**Comment by vishwassathish** [comment]
Thank you for the clarification! Appreciate your time and efforts to answer my questions. My issue with exploding gradients turned out to be related to incorrect `train_batch_size` and `ppo_min_batch_size`. Observing more stable training now. I will close this issue for now.
